### DISCUSSION TITLE
Should completions work when using python -m pipx?

### DISCUSSION THREAD

**Discussion by doolio** [pypa/pipx#1376] [Q&A]
They don't but I hoped they would.

**Comment by SirHegel** [comment]
They can, but not through the registration in `pipx completions` — that one is bound to the word `pipx`, and the shell picks a completion by the **first** word on the line.

`eval "$(register-python-argcomplete pipx)"` ends with:

```bash
complete -o nospace -o default -o bashdefault -F _python_argcomplete pipx
```

So when you type `python -m pipx <TAB>`, bash looks for a completion registered for `python`, finds none, and falls back to filename completion. Registering `_python_argcomplete` for `python` instead is not a fix — it would hijack completion for every `python` invocation.

The module itself is fine. `python -m pipx` answers argcomplete's protocol correctly; you can call it by hand:

```console
$ _ARGCOMPLETE=1 _ARGCOMPLETE_IFS=$'\n' _ARGCOMPLETE_SHELL=bash \
  COMP_LINE="pipx ins" COMP_POINT=8 COMP_TYPE=9 \
  python -m pipx 8>&1 1>/dev/null
install
install-all
```

So all that is missing is a first word the shell can key on. A function is enough:

```bash
pipx() { python -m pipx "$@"; }
eval "$(register-python-argcomplete pipx)"
```

Tested — with that in place, `pipx ins<TAB>` gives `install install-all`, same as a real `pipx` on PATH. Put both lines in your shell config, in that order.

One thing to avoid: `register-python-argcomplete -e "python -m pipx" pipx` looks like the right tool, but with a multi-word command it emits

```bash
compdef _python_argcomplete_/usr/bin/python_SPACE_-m_SPACE_pipx pipx
```

and a `/` is not valid in a shell function name, so sourcing it silently defines nothing. `-e` wants a single executable path.

<sub>Verified against argcomplete's generated script and by running the protocol above rather than from memory; Claude helped me dig through `register-python-argcomplete`'s output.</sub>
